ramips: Add support for Creality WB-01

Npeca75 opened the pull request:
CREALITY BOX WB01 is small footprint router based on MediaTek MT7688

Specifications:
- SoC: MediaTek MT7688 580MHz
- RAM: DDR2 128M
- Flash: BY25Q128AS (16 MiB, SPI NOR) handled by BoHong bh25q128as driver
- WiFi: 2.4GHz 1T1R internal panel antenna
- Ethernet: 1x LAN (10/100)
- USB: 2x USB2.0 port
- UART: 3.3V, TX, RX, GND / 56700 8N1 / only pads on PCB
- micro USB input (for power only)
- reset button
- FCC ID: 2AXH6CREALITY-BOX

MAC addresses as verified by OEM firmware:
```
vendor   OpenWrt    source
LAN      eth0       factory 0x2e
2.4GHz   phy0-ap0   factory 0x04 (label)
```

LEDs
```
color    vendor                OpenWRT     configurable
red      SD card activity        -             yes
green    Claoud connectivity   status          yes
blue     LAN activity          eth0            yes
yellow   WIFI activity         phy0tpt         yes
```

Return to OEM & debrick
- download "cxsw_update.tar.bz2" from manufacturer site
- extract archive to FAT32 USB stick root
- put USB stick in USB2 port
- press & hold reset button
- power on device while holding reset
- wait approx 10 sec
- release reset button

Installation with SD Card
- power on device
- wait for device to finish starting
- copy "openwrt-ramips-mt76x8-creality_wb-01-squashfs-cxsw_update.tar.bz2" to root of FAT32 SD card
- rename openwrt-ramips-mt76x8-creality_wb-01-squashfs-cxsw_update.tar.bz2 to "cxsw_update.tar.bz2"
- put SD card in device
- divice will install OpenWRT on internal flash

Installation via telnet:
- extract the "factory.bin" and "install.sh" from newly created openwrt-ramips-mt76x8-creality_wb-01-squashfs-cxsw_update.tar.bz2 to FAT32 USB stick root
- telnet to 10.10.10.254, user: root, password: cxswprin
- plug the USB in USB1 port
- cd /media/usbdisk/
- sh install.sh
- device will write "factory.bin" to internal flash

rmilecki commented on target/linux/ramips/dts/mt7628an_creality_wb-01.dts:
```diff
@@ -0,0 +1,138 @@
+		compatible = "gpio-leds";
+
+		wlan2g {
+			label = "yellow:wlan2g";
```
Please use `function` (`LED_FUNCTION_WLAN_2GHZ`) and `color` (`LED_COLOR_ID_YELLOW`) instead of `label`.

Applies to other LEDS as well.

Npeca75 replied:
hi @rmilecki 
i was hoping that this device could be backported to 23.05, but with this new LED semantic ..
anyway ... done as you suggested :)
hope it is OK now

shivajiva101 commented:
Duplicate of https://github.com/openwrt/openwrt/pull/3802

Npeca75 commented:
> Duplicate of #3802 You should at least credit the original author...

hi @shivajiva101 
tnx for notification
there is 2 PR for this device, but both are unfinished and abandoned and somewhat wrong (MAC address, ETH switch)
#13150 and #3802 
would you be so kind to guide me how and where to mention other two author?
this is my first PR so any help is welcome :)

shivajiva101 commented:
Hi @Npeca75 chepo92 made it clear that it was the work of figgc he was effectively bumping by including him as the author in the pr commit message, you should probably do the same just so there's no confusion as to the source. See https://github.com/openwrt/openwrt/pull/13150#issuecomment-1651885589 for how to add yourself properly to the chain.

Npeca75 commented:
> you should probably do the same

hi @shivajiva101 
since i started my work on this device without knowing that these PRs existed, until @andyboeh pointed out in forum :slightly_smiling_face: i could disagree with you ... but, my goal is to add this device to OpenWRT and not to argue :slightly_smiling_face: 

so, i am tried to amend my first commit to add author and as you wished myself as co-author, but .. looks like i have limited knowledge of git

what i tried is
git rebase -i a6991fc7d251f7ab65a588546e22abfd4f8ce472 (commit before mine)
rewording
added author & co-author
git push --force

but, looks like i messed up my branch somehow
do you have any idea how to do it properly ?
i have a backup of my branch if it is needed

hauke commented:
Please rebase, it loos like your last push broke something

Npeca75 commented:
> Please rebase, it loos like your last push broke something

thank you :)

shivajiva101 commented:
It's whatever you're comfortable with at the end of the day. Personally I'd like to see the hardware officially supported, it will make my life easier :man_shrugging:

Npeca75 commented:
> at the end of the day

well, i restored my original branch after all
CI/CD complaining about formalities with modified author, sorry, i have no more spare time for endless try

shivajiva101 commented:
that's convenient for you, along with the claim it was clean room coded despite using the same older code style as figgyc and needing to be asked to update the leds format. I'm surprised the devs aren't being more vocal on how this appears but hey ho let's hope you can get it merged for everyone's convenience

Npeca75 commented:
> that's convenient for you, along with the claim it was clean room coded despite using the same older code style as figgyc and needing to be asked to update the leds format. I'm surprised the devs aren't being more vocal on how this appears but hey ho let's hope you can get it merged for everyone's convenience

ok, since your opinion is that every similarity in LED naming, or in ? in what? install procedure? is pure robbery ? steal ? should i close this PR for your satisfy ?

chepo92 commented:
@Npeca75 great job and nice to have a more detailed commit can we colaborate so we get any pr merged? 
As you said 'CI/CD complaining about formalities with modified author, sorry, i have no more spare time for endless try' 
Well, I did invest some time (I would said a lot) to pass CI/CD, was a PITA, but then got stuck with reviewer's comment and "didn't have more time" 

Besides, Shivajiva and other guys have been doing some further work in the discord, you should join if interested

Npeca75 commented:
> can we colaborate

hi @chepo92 , yes, i am open to anything which will bring this device to official support ... but please, could we message each other on OWRT forum about collaboration? i am not so good with GIT and since it is my first PR , mostly confused with workflow ...

shivajiva101 commented:
> > that's convenient for you, along with the claim it was clean room coded despite using the same older code style as figgyc and needing to be asked to update the leds format. I'm surprised the devs aren't being more vocal on how this appears but hey ho let's hope you can get it merged for everyone's convenience
> 
> ok, since your opinion is that every similarity in LED naming, or in ? in what? install procedure? is pure robbery ? steal ? should i close this PR for your satisfy ?

Don't be silly I want the PR merged. Just pointing out someone already bought the hardware, peeled it open and did the reverse engineering to make the first PR for this, regardless of you claiming ignorance to that event. Seems logical that they would be credited in any subsequent PR out of respect for their efforts, once you were aware, and that's literally all I'm pointing out to you. A major portion of the code for your PR was already in existence in the first PR due to the nature of the structure created specifically for adding hardware.

Npeca75 commented:
> Just pointing out someone already bought the hardware, peeled it open

same as i did. it is here, attached on 2pcs of USB printer

since p910nd was broken for 2 usb port operation, i sent a PR to packages which is accepted, since then p910nd is OK
and, yes, this is why i sent PR for this device, without knowing for other two PRs
if you take a closer look, i am using switch, other two is using plain ETH0
image generating script is again, similar, but not same
in other two PR there is no option for flash with telnet ... so much about relying on other people job
my only sin is that i made same image name as in other PR since it is logical to be similar to "factory update file"

anyway, if you wish to help me, please help me trough GIT how to done it properly

shivajiva101 commented:
Sure, as the owner of the pr you can edit your commit comments, make further changes etc up to the merge point where it will probably be squashed into a single commit. You don't have to force push anything just keep working on the branch the pr originates from. Make the changes requested by the dev handling this and keep on it so it doesn't sit forever.

andyboeh commented:
> anyway, if you wish to help me, please help me trough GIT how to done it properly

What's your problem with git?

Usually, I use the GUI (gitk) for most operations (rebasing is an exception). I create a single commit with my changes. If the review requires changes (that is very, very common), then I make the changes and select "Commit" -> "Amend last commit" to just get it into the same commit. Force-Push and you're done. No need to squash or merge anything.

That said, for rebasing, I follow two guides in this order:

1. Sync my fork according to the Github tutorial at https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork#syncing-a-fork-branch-from-the-command-line
2. Rebase according to the Git book at https://git-scm.com/book/en/v2/Git-Branching-Rebasing

@shivajiva101 With my first few PRs, I was always told to squash everything into a single commit. I wasn't very proficient with git at that time, so I established the above-mentioned workflow.

shivajiva101 commented:
@andyboeh fair enough, the pr history isn't relevant to the project history tree and the requested changes are visible in the pr regardless, so it's a valid approach and definitely looks cleaner

shivajiva101 commented:
I think the issue with the Test Formalities check failing is trying to author figgyc in when you are the author of the pr, github doesn't allow it to prevent forged commits iirc

Npeca75 commented:
> What's your problem with git?

i was confused where and how should i mention figgyc in commit, and since @shivajiva101 was demanding to include figgyc, i was hopping that he will give me some real and usable instruction how to do this 
... my first try with changing initial commit message was a mess, but hope now is OK

shivajiva101 commented:
@Npeca75 your last push looks okay to me :+1:

Npeca75 commented:
@shivajiva101 
> Don't be silly I want the PR merged.
>your last push looks okay to me

would you please review and approve this PR ?

@rmilecki 
may i ask you to finish review of this PR ?

shivajiva101 commented:
@Npeca75 I don't have write access so I can't be of any help to you in that respect. Can the blue led be configured as link on, transmit & receive to reflect the status on that interface as the default behaviour?

shivajiva101 commented:
@rmilecki  As this seems to have stalled, could you express what is missing from the pr to get it moving along pls?

hauke commented:
Please rebase on top of current main branch and solve the conflicts.

Npeca75 committed:
ramips: CREALITY BOX WB01 is small footprint router based on MediaTek MT7688

Specifications:
- SoC: MediaTek MT7688 580MHz
- RAM: DDR2 128M
- Flash: BY25Q128AS (16 MiB, SPI NOR) handled by BoHong bh25q128as driver
- WiFi: 2.4GHz 1T1R internal panel antenna
- Ethernet: 1x LAN (10/100)
- USB: 2x USB2.0 port
- UART: 3.3V, TX, RX, GND / 56700 8N1 / only pads on PCB
- micro USB input (for power only)
- reset button
- FCC ID: 2AXH6CREALITY-BOX

MAC addresses as verified by OEM firmware:
vendor   OpenWrt    source
LAN      eth0       factory 0x2e
2.4GHz   phy0-ap0   factory 0x04 (label)

LEDs
color    vendor                OpenWRT     configurable
red      SD card activity        -             yes
green    Cloud connectivity    status          yes
blue     LAN activity          eth0            yes
yellow   WIFI activity         phy0tpt         yes

Return to OEM & debrick
- download "cxsw_update.tar.bz2" from manufacturer site
- extract archive to FAT32 USB stick root
- put USB stick in USB2 port
- press & hold reset button
- power on device while holding reset
- wait approx 10 sec
- release reset button

Installation with SD Card
- power on device
- wait for device to finish starting
- copy "openwrt-ramips-mt76x8-creality_wb-01-squashfs-cxsw_update.tar.bz2"
  to root of FAT32 SD card
- rename openwrt-ramips-mt76x8-creality_wb-01-squashfs-cxsw_update.tar.bz2
  to "cxsw_update.tar.bz2"
- put SD card in device
- device will install OpenWRT on internal flash

Installation via telnet:
- extract the "factory.bin" and "install.sh" from newly created
  openwrt-ramips-mt76x8-creality_wb-01-squashfs-cxsw_update.tar.bz2
  to FAT32 USB stick root
- telnet to 10.10.10.254, user: root, password: cxswprin
- plug the USB in USB1 port
- cd /media/usbdisk/
- sh install.sh
- device will write "factory.bin" to internal flash

Co-authored-by: George Brooke <figgyc@figgyc.uk>
Signed-off-by: Peca Nesovanovic <peca.nesovanovic@sattrakt.com>

Npeca75 commented:
> Please rebase on top of current main branch and solve the conflicts.

hi @hauke 
done !

shivajiva101 commented:
Just built this and there is still an issue with v24, it's the same problem when using figgy's patch...
``` 
Wed Feb 19 17:21:07 2025 kern.err kernel: [   10.822322] mmc0: Card stuck being busy! __mmc_poll_for_busy
Wed Feb 19 17:21:07 2025 kern.warn kernel: [   11.120083] mtk-sd 10130000.mmc: no support for card's volts
Wed Feb 19 17:21:07 2025 kern.err kernel: [   11.131432] mmc0: error -22 whilst initialising SDIO card 
```
It can't access the mmc device, maybe related to #17446

DragonBluep commented on target/linux/ramips/dts/mt7628an_creality_wb-01.dts:
```diff
@@ -0,0 +1,143 @@
+
+&wmac {
+	status = "okay";
+	mediatek,mtd-eeprom = <&factory 0x0>;
```
Use NVMEM.

DragonBluep commented on target/linux/ramips/dts/mt7628an_creality_wb-01.dts:
```diff
@@ -0,0 +1,143 @@
+
+&sdhci {
+	status = "okay";
+	mediatek,cd-low;
```
Please add the correct pinctrl to fix the SD card issue. https://github.com/openwrt/openwrt/pull/17446

DragonBluep commented on target/linux/ramips/dts/mt7628an_creality_wb-01.dts:
```diff
@@ -0,0 +1,143 @@
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	macaddr_factory_04: macaddr@04 {
```
Squash this block into `factory` partition.
```suggestion
	macaddr_factory_4: macaddr@4 {
```

DragonBluep commented on target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:
```diff
@@ -173,6 +173,7 @@ ramips_setup_interfaces()
 		ucidef_set_interface "wan" device "/dev/cdc-wdm0" protocol "qmi"
 		;;
+	creality,wb-01|\
 	tplink,tl-mr3020-v3)
```
Sort this block alphabetically.

chepo92 commented:
@Npeca75 Hi, are you able to make the suggested changes to the PR to get it finally merged, maybe @shivajiva101 or I can help

hauke commented:
Support for the device was merged in f29bc8736aacf36fc4f339182b74a20489096a2b using #19686.
